### DISCUSSION TITLE
`FuturesUnordered` and `yield_now()`

### DISCUSSION THREAD

**Discussion by duarten** [tokio-rs/tokio#6090] [Q&A]
`FuturesUnordered` [detects](https://docs.rs/futures-util/0.3.26/src/futures_util/stream/futures_unordered/mod.rs.html#527) whether the futures it is managing intend to yield by checking whether the waker has been notified when the poll result was `Pending`. If there are two instances of this, `FuturesUnordered` will itself yield.

However, Tokio's `yield_now` utility doesn't follow this pattern, as it defers so the waker is only notified after the scheduler runs the I/O loop (from what I could tell). In [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a895637b6350f752969f07eb55c7f727) you can see how these two behaviors of yielding change the schedule order.

It also seems that Tokio has a [similar detection](https://github.com/tokio-rs/tokio/blob/881b510a072f5acd773a10d3be0debf74113404e/tokio/src/runtime/task/harness.rs#L154), so I was wondering if it makes sense to change Tokio's `yield_now` to immediately wake the waker?

**Comment by Darksonn** [comment]
We used to do that in the past, but we changed it in #5223.

**Reply by duarten** [reply]
Is there a different way to implement https://github.com/tokio-rs/tokio/pull/5223, or is the situation just hopeless until something like [this proposal](https://internals.rust-lang.org/t/runtime-agnostic-cooperative-task-scheduling-budget/18796/40) comes to be?
